src/modules/python/pbs_hooks/PBS_alps_inventory_check.PY

```diff
@@ -0,0 +1,348 @@
+    __exit_hook(0, msg)
+
+# XTHOSTNAME file found on this host. Read it to determine our Cray hostname.
+
+with open(XTHOSTNAME) as xthost_file:
+    my_crayhost = xthost_file.readline()
+    my_crayhost = my_crayhost.rstrip()
+    msg += ["Processing ALPS inventory for crayhost %s" % (my_crayhost)]
+
+start = time.time()
+vnodes = pbs.server().vnodes()
+vnodes_query_duration = time.time() - start
+if not vnodes:
+    msg += ["ALPS Inventory Check: No vnodes reported by PBS"]
+    __exit_hook(1, msg)
+
+down_states = pbs.ND_DOWN | pbs.ND_STALE | pbs.ND_STATE_UNKNOWN
+
+for v in vnodes:
+    str_v = str(v)
+    vntype = " "
+
+    if (v.state & down_states) or \
+        "PBScrayhost" not in v.resources_available or \
+            v.resources_available["PBScrayhost"] != my_crayhost:
+        continue
+
+    if "vntype" in v.resources_available:
+        vntype = v.resources_available["vntype"]
+
+    if vntype == "cray_login":
+        if (not (v.state & pbs.ND_OFFLINE)) and \
+                str_v not in cray_login_list:
+            cray_login_list.append(str_v)
+
+    elif vntype == "cray_compute":
+
+        if "PBScraynid" in v.resources_available:
+            pbs_craynid = v.resources_available["PBScraynid"]
+
+        if pbs_craynid.isdigit():
+            if (v.state & pbs.ND_OFFLINE):
+
+                # if vnode is offline, add it to offline_nids_list.
+                # Otherwise, add it to the pbs_nids_set.
+                # Later on, the inventory MoM sill iterate through the
+                # offline_nids_list and discard those nids from the
+                # apstat_nids_set. This has the effect of causing us
+                # to ignore any offline nids so we won't generate a
+                # spurious HUP of the MoM if there are offline vnodes
+                # whose nids are present in the apstat output. We use
+                # the set.discard() method because it doesn't throw an
+                # error if the nid isn't present in the apstat_nids_set.
+
+                offline_nids_list.append(int(pbs_craynid))
+            else:
+                pbs_nids_set.add(int(pbs_craynid))
+
+if len(cray_login_list) == 0:
+    msg += ["ALPS Inventory Check: No eligible " +
+            "login nodes to perform inventory check"]
+    __exit_hook(0, msg)
+
+cray_login_local_name = pbs.get_local_nodename()
+
+try:
+    inventory_node = cray_login_list[0]
+    inventory_addr = socket.gethostbyname(inventory_node)
+
+    if ((inventory_addr not in my_addresses()) and (
+            inventory_node != cray_login_local_name)):
+        msg += ["ALPS Inventory Check: Login node '%s' is in charge of "
+                "verification, skipping check on '%s'." %
+                (inventory_node, socket.gethostname())]
+        __exit_hook(0, msg)
+
+    start = time.time()
+    apstat_nids_set = get_apstat_nids(msg)
+    apstat_query_duration = time.time() - start
+    if apstat_nids_set is None or len(apstat_nids_set) == 0:

```

Comment:
This check should be moved to get_apstat_nids(). We should exit only if apstat command failed.